Thrust limiter

jiena76 opened the pull request.

jiena76 committed:
first commit wrote thrust limiter class with max ramping speed, pending design change

jiena76 committed:
change algorithm to the available power based function, change the number of thrust values 8 -> 6, speed of thrust values will be scaled so that they're under the power cap, test cases not adjusted yet

jiena76 committed:
change file name from Thrust_Limiter to Force_Limiter and class name to ThrustLimiter to ForceLimiter

jiena76 committed:
update code using: https://purduerov.slack.com/files/U0DFSVAMC/F8N39KQ78/limiting_notes.py, function "correct_thrust_vector" -> "enforce", delete self.speed, add self.a6vector, an8vector, and last8vector, make "acceptable_individual_change" and "acceptable_total_change" function to intake "change" vector as a parameter (instead of declaring that variable in each function), update last8vector at the end of "enforce" function, add TODOs for each function

jiena76 committed:
add four constants: POWER_THRESH, IND_THRESH, TOTAL_CHANGE_THRESH, and IND_CHANGE_THRESH

jiena76 committed:
add a new line at the end of the file

jiena76 committed:
change all constants names to all capital letters

jiena76 committed:
change function "enforce"'s function calling orders to 1. acceptable_total_power, 2. acceptable_individual_total, 3. acceptable_total_change, 4. acceptable_individual_change

bmaxfie committed:
Thrust Limiter code reviewed by Ben.

bmaxfie committed:
Added disabledT parameter to ForceLimiter enforce(). Will need for mappers.

jiena76 committed:
'change' variable -> difference8

jiena76 committed:
add convertToPowerUnit() method

jiena76 committed:
'change' variable -> difference8 (missed one in the previous commit)
add the equation for convertToPowerUnit as a comment line

mattmolo commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,117 @@
+    def __init__(self):
+        self.a6vector = [0 for _ in range(6)]
+        self.an8vector = [0 for _ in range(8)]
+        self.last8vector = [0 for _ in range(8)]
```
I don't quite understand the naming here. Can you make more descriptive names? I think they correspond to something like this:
`a6vector = output_force`
`an8vector = thruster_forces`
`last8vector = last_thruster_forces`

bmaxfie replied:
Agreed

jiena76 replied:
update:
a6vector = force_direction
an8vector = thruster_powers
last8vector = prev_thruster_powers
difference8 = diff_thruster_powers

mattmolo commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,117 @@
+    1. acceptable_total_power
+    2. acceptable_individual_total
+    3. acceptable_total_change
+    4. acceptable_individual_change
```
I think theses are the constants? I think it'd be good to explain what they mean (and there units if they have them), what the correspond to in `init_hw_constants` and why they're needed.

bmaxfie replied:
they are not constants. They are the limiting functions. Constants are within them, but will be externalized to init_hw_constants when we finish figuring out what they need to be.

mattmolo replied:
Oh, derp. I see. Okay, well that's fine I guess.

jiena76 replied:
added "Order of Limiting Functions" comment

mattmolo commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,117 @@
+    def acceptable_individual_total(self):
+        # tests whether each individual thruster is exceeding the maximum
+        # "Limiting"
+        # should return 1 if everything is acceptable, otherwise a value \in (0,1) which will be multiplied to the 6vector
```
~\in~ => in? Just small typo I think you have here and in the other places

jiena76 replied:
fixed typos

mattmolo commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,117 @@
+    def enforce(self, a6vector, last8vector, disabledT):
+
+        self.a6vector = a6vector
+        self.an8vector = mapper.calculate(a6vector, disabledT)
```
I feel like the `mapper.calculate` should be done outside the enforce calculation and passed into you. I guess this depends on how the whole thruster control system is written.

bmaxfie replied:
True mapping is supposed to be done after thrust limiting. So its done within here just so we can get their temporary values.

mattmolo replied:
I thought the process is:
1) True Mapping
2) Check thrust limiting
3) Readjust forces if not acceptable, go back to step 1
4) Yay, we're done

bmaxfie replied:
I mean, what's the difference? It's more linear and easy to understand on the high level the current way.

mattmolo replied:
Well wait, so you're running thrust mapping in here, modifying the force values, returning the force values, then doing thrust mapping again outside this function?

mattmolo commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,117 @@
+        if ind_change_scale != 1:
+            self.a6vector = ind_change_scale * self.a6vector
+
+        return self.a6vector
```
Since you're returning 1.0 when checking all the power and change levels, a cleaner way to do this might just be:
```
return (self.a6vector * self.acceptable_total_power()
                      * self.acceptable_individiual_total()
                      * self.acceptable_total_change(difference8) 
                      * self.acceptable_individual_change(difference8))
```

bmaxfie replied:
Agreed

jiena76 replied:
applied

mattmolo commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,117 @@
+        IND_CHANGE_THRESH = 1000
+
+        # get greatest absolute value
+        max_change = abs(max(min(difference8), max(difference8), key=abs))
```
Could this be `max(abs(difference8))`?

bmaxfie replied:
Agreed

jiena76 replied:
applied

mattmolo reviewed (commented):
Hi there, I was going through all the pull requests and taking a look at everything. Just wanted to give a small review. Let me know if I misunderstood anything or if you think I'm wrong!

jiena76 committed:
change
a6vector = force_direction
an8vector = thruster_powers
last8vector = prev_thruster_powers
difference8 = diff_thruster_powers
edit few typos in comments

jiena76 committed:
change
abs(max(min(difference8), max(difference8), key=abs))
to
max(abs(difference8))
and change enforce() limiting function calls to
return (self.a6vector * self.acceptable_total_power()
                      * self.acceptable_individiual_total()
                      * self.acceptable_total_change(difference8)
                      * self.acceptable_individual_change(difference8))
this format

bmaxfie commented:
Okay, @jiena76's work on this is complete. The next step is adding real constants and conversion functions by Lindsey. I'm not merging this pull request until the code is fully functioning.

mattmolo commented:
thanks for all the changes @jiena76 😄

Commit:
Rewrote thrust mapper and combined with thrust limiting

Commit:
Merge branch 'thrust_limiter' of https://github.com/purduerov/XX-Core into thrust_limiter

Commit:
Completed total power calculations and set up individual thruster power limiting. Not complete or runnable yet, more work needed to ensure thrust ratios are the same with thrust limiting

Commit:
Completed draft of thrust mapping and limiting code, testing needed and possible minor adjustments

Commit:
Added comments at beginnning of code to say which thrusters are described by which positions in code matrices

Commit:
Added boolean to disable thrust limiting to calculate function

bmaxfie reviewed (changes requested):
A few things to address...

bmaxfie commented on rov/init_hw_constants.py:
```diff
@@ -3,3 +3,7 @@
 NEG_MAX_POWER = 222  # Value to set thrusters to max negative power
 POS_MAX_POWER = 388  # Value to set thrusters to max positive power
 FREQUENCY = 47  # Frequency at which the i2c to pwm will be updated
+POWER_THRESH = 1000 # arbitrary power threshold value
```
I thought you had the actual values for these thresholds now? If so, change the comments to show they aren't 'arbitrary'

bmaxfie commented on rov/movement/mapper/Complex_1.py:
```diff
@@ -0,0 +1,275 @@
+    thrust map matrix = pseudo-inverse(A) * desired thrust
+    *For location and rotation vectors, the first four thrusters are the horizontal thrusters: first is front left, 
+    second is front right, third is back left, and fourth is back right. The last four are the vertical thrusters in
+    the same order: front left, front right, back left, and back right.
```
Explain at the end, how outside classes should use this class. What functions to call to do what, when, etc.

bmaxfie commented on rov/movement/mapper/Complex_1.py:
```diff
@@ -0,0 +1,275 @@
+        [0, 0, 0, 0, 1, 1, 1, 1]
+    ])
+	
+    THRUST = np.matrix([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
```
Convention usually says capital variables are constants. Review all class fields pls

bmaxfie commented on rov/movement/mapper/Complex_1.py:
```diff
@@ -0,0 +1,275 @@
+        if max_val > 1:
+            self.map /= max_val
+
+    def limit_power(self, initialPower):
```
What helps with outside users understanding how to use Python classes is identifying them as 'private' if they are only supposed to be used by the class. Python convention is to begin variables and functions with an underscore ( _ ).

bmaxfie commented on rov/movement/thruster/Force_Limiter.py:
```diff
@@ -0,0 +1,105 @@
+from math import floor
```
If this class is not used anymore, delete it. We can always get it again through git anyway.

bmaxfie reviewed (changes requested):
Okay, these are the real changes ... O.o

bmaxfie commented:
@lindseykmay7

Commit:
Updated code to meet standards, no logic changes

Commit:
Changes power threshold values to be used in Complex_1.py

Commit:
Deleted because no longer being used

lindseykmay7 commented:
Code has been updated to reflect requested changes and is ready to be merged.
